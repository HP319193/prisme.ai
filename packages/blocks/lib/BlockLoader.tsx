import './i18n';
import * as React from 'react';
import { ReactElement, ReactNode, useState } from 'react';
import { useTranslation } from 'react-i18next';
import { BlockProvider, BlockProviderProps } from './Provider';
import { useBlocks } from './Provider/blocksContext';
import * as builtinBlocks from './Blocks';
import { Schema } from '@prisme.ai/design-system';
import useExternalModule from './utils/useExternalModule';

class BlockErrorBoundary extends React.Component<{ children: ReactElement }> {
  state = {
    hasError: false,
  };

  static getDerivedStateFromError() {
    return { hasError: true };
  }

  componentDidCatch(error: any, errorInfo: any) {
    // Vous pouvez aussi enregistrer l'erreur au sein d'un service de rapport.console.error(error, errorInfo);
  }

  render() {
    if (this.state.hasError) {
      // Vous pouvez afficher n'importe quelle UI de repli.
      return <h1>Something went wrong.</h1>;
    }

    return this.props.children;
  }
}

export interface BlockComponentProps {
  token?: string;
  workspaceId?: string;
  appInstance?: string;
  language?: string;
  edit?: boolean;
  layout?: {
    container?: HTMLElement;
  };
}
export type BlockComponent = (
  props: BlockComponentProps
) => ReactElement & {
  schema?: Schema;
};

export interface BlockLoaderProps extends BlockComponentProps {
  children?: ReactNode;
  url?: string;
  name?: string;
  onLoad?: (block: any) => void;
}

export const ReactBlock = ({
  url = '',
  onLoad,
  ...componentProps
}: BlockLoaderProps) => {
  const {
    i18n: { language },
  } = useTranslation('workspaces');
  const {
    externals,
    components: { Loading },
  } = useBlocks();

  const { module: Component, loading } = useExternalModule({
    url,
    externals,
  });

  React.useEffect(() => {
    if (!onLoad || !module || loading) return;
    onLoad(module);
  }, [onLoad, module, loading]);

  return (
    <>
      {loading && componentProps.edit && <Loading />}
      {Component && <Component {...componentProps} language={language} />}
    </>
  );
};

export const IFrameBlock = ({ url, token, edit }: BlockLoaderProps) => {
  const [loading, setLoading] = useState(true);
  const [height, setHeight] = useState(100);
  const {
    components: { Loading },
  } = useBlocks();
  const handleLoad = React.useCallback(
    (e: any) => {
      setLoading(false);
      e.target.contentWindow.postMessage(
        {
          source: 'prisme.ai',
          token,
        },
        '*'
      );
    },
    [token]
  );

  return (
    <>
      {loading && edit && <Loading />}
      <iframe
        src={url}
        onLoad={handleLoad}
        className="flex"
        style={{
          width: '100%',
          height: `${height}px`,
        }}
      />
    </>
  );
};

const getComponentByName = (name: string) =>
  builtinBlocks[name as keyof typeof builtinBlocks] || null;

const BlockRenderMethod = ({ name, url, ...props }: BlockLoaderProps) => {
  if (name) {
    const Component = getComponentByName(name);
    if (Component) {
      return <Component edit={!!props.edit} {...props} />;
    }
  }

  const isJs = url && url.replace(/\?.*$/, '').match(/\.js$/);
  if (isJs) {
    return <ReactBlock url={url} {...props} />;
  }

  return <IFrameBlock url={url} {...props} />;
};

export const BlockLoader = ({
  config,
  onConfigUpdate,
  onAppConfigUpdate,
  api,
  ...props
}: BlockProviderProps & BlockLoaderProps) => {
  return (
    <BlockErrorBoundary>
      <BlockProvider
        config={config}
        onConfigUpdate={onConfigUpdate}
        appConfig={props.appConfig}
        onAppConfigUpdate={onAppConfigUpdate}
        events={props.events}
        api={api}
        onLoad={props.onLoad}
      >
        <BlockRenderMethod {...props} />
      </BlockProvider>
    </BlockErrorBoundary>
  );
};
