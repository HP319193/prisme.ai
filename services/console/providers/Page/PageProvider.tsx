import { FileUnknownOutlined } from '@ant-design/icons';
import { Loading } from '@prisme.ai/design-system';
import { useTranslation } from 'next-i18next';
import {
  createContext,
  ReactNode,
  useEffect,
  useCallback,
  useState,
} from 'react';
import NotFound from '../../components/NotFound';
import api from '../../utils/api';
import { useContext } from '../../utils/useContext';

interface Page extends Prismeai.Page {
  apiKey?: string;
}

export interface PageContext {
  page: Page;
  appInstances: Prismeai.PageDetails['appInstances'];
  loading: boolean;
  fetchPage: () => Promise<Page | null>;
  savePage: (page: Page) => Promise<Page | null>;
  saving: boolean;
  deletePage: () => Promise<Page | null>;
}

export const pageContext = createContext<PageContext | undefined>(undefined);

export const usePage = () => useContext<PageContext>(pageContext);

interface PageProviderProps {
  workspaceId?: string;
  slug?: string;
  children: ReactNode;
}

export const PageProvider = ({
  workspaceId,
  slug,
  children,
}: PageProviderProps) => {
  const { t } = useTranslation('workspaces');
  const [page, setPage] = useState<PageContext['page']>();
  const [appInstances, setAppInstances] = useState<PageContext['appInstances']>(
    []
  );
  const [loading, setLoading] = useState<PageContext['loading']>(true);
  const [saving, setSaving] = useState<PageContext['saving']>(false);
  const [notFound, setNotFound] = useState(false);

  const fetchPage = useCallback(async () => {
    setNotFound(false);
    if (!workspaceId || !slug) return null;
    try {
      const { appInstances, public: isPublic, ...page } = await api.getPage(
        workspaceId,
        slug
      );
      setAppInstances(appInstances);
      return page || null;
    } catch (e) {
      setNotFound(true);
      return null;
    }
  }, [slug, workspaceId]);

  const savePage: PageContext['savePage'] = useCallback(
    async ({ apiKey, ...newPage }) => {
      if (!workspaceId) return null;
      setSaving(true);
      const updated = await api.updatePage(
        workspaceId,
        newPage,
        page && page.slug
      );
      setPage(updated);
      setSaving(false);
      return updated;
    },
    [page, workspaceId]
  );

  const deletePage: PageContext['deletePage'] = useCallback(async () => {
    if (!workspaceId || !page?.slug) return null;
    setPage(undefined);
    api.deletePage(workspaceId, page.slug);
    return page;
  }, [page, workspaceId]);

  useEffect(() => {
    const initPage = async () => {
      setLoading(true);
      const page = await fetchPage();
      setLoading(false);
      if (!page) return;
      setPage(page);
    };
    initPage();
  }, [fetchPage, slug]);

  if (loading) return <Loading />;
  if (notFound)
    return <NotFound icon={FileUnknownOutlined} text={t('pages.notFound')} />;
  if (!page) return null;

  return (
    <pageContext.Provider
      value={{
        page,
        appInstances,
        loading,
        fetchPage,
        savePage,
        saving,
        deletePage,
      }}
    >
      {children}
    </pageContext.Provider>
  );
};

export default PageProvider;
