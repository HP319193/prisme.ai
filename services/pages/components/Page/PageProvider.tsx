import { createContext, FC, useContext } from 'react';
import usePageEvents from './usePageEvents';
import usePageFetcher from './usePageFetcher';

export interface PageContext
  extends ReturnType<typeof usePageFetcher>,
    ReturnType<typeof usePageEvents> {}
export const pageContext = createContext<PageContext>({
  page: null,
  setPage() {},
  loading: false,
  async fetchPage() {},
  blocksConfigs: [],
  events: undefined,
});

export const usePage = () => useContext(pageContext);

export const PageProvider: FC = ({ children }) => {
  const { page, setPage, loading, fetchPage } = usePageFetcher();
  const { blocksConfigs, events } = usePageEvents(page);

  return (
    <pageContext.Provider
      value={{ page, setPage, loading, fetchPage, blocksConfigs, events }}
    >
      {children}
    </pageContext.Provider>
  );
};

export default pageContext;
