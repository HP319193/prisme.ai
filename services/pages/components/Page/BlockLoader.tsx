import { BlockLoader as BLoader, TBlockLoader } from '@prisme.ai/blocks';
import { useTranslation } from 'next-i18next';
import { useCallback, useEffect, useRef, useState } from 'react';
import { useUser } from '../../../console/components/UserProvider';
import api from '../../../console/utils/api';
import Storage from '../../../console/utils/Storage';
import { usePage } from './PageProvider';
import { useDebug } from './useDebug';

/**
 * This function aims to replace deprecated Block names by the new one
 */
function getBlockName(name: string) {
  switch (name) {
    case 'Layout':
      return 'StackedNavigation';
    default:
      return name;
  }
}

export const BlockLoader: TBlockLoader = ({
  name = '',
  config: initialConfig,
  onLoad,
  container,
}) => {
  const { user } = useUser();
  const [url, setUrl] = useState('');
  const [config, setConfig] = useState<typeof initialConfig>(initialConfig);
  const [appConfig, setAppConfig] = useState<any>();
  const { page, events } = usePage();
  const [loaded, setLoaded] = useState(false);
  const {
    i18n: { language },
  } = useTranslation();
  const lock = useRef(false);
  const [listening, setListening] = useState(false);

  const debug = useDebug();

  useEffect(() => {
    setConfig(initialConfig);
  }, [initialConfig]);

  useEffect(() => {
    if (lock.current || !name) return;
    if (name.match(/^http/)) {
      setUrl(name);
      return;
    }

    const parts = name.split(/\./);

    if (parts.length === 1) {
      const { blocks: workspaceBlocks } =
        (page?.appInstances || []).find(({ slug }) => slug === '') || {};
      if (workspaceBlocks && workspaceBlocks[name]) {
        setUrl(workspaceBlocks[name]);
        return;
      }
      setUrl(getBlockName(name));
      return;
    }

    const [appSlug] = parts;
    const app = (page?.appInstances || []).find(({ slug }) => appSlug === slug);
    if (!app || !app.blocks?.[name]) {
      console.error(`"${name}" Block is not installed`);
      return;
    }

    const debugUrl = debug.get(name);
    if (debugUrl) {
      app.blocks[name] = debugUrl;
    }
    setAppConfig(app.appConfig);
    setUrl(app.blocks[name]);
  }, [debug, name, page]);

  const { onInit, updateOn, automation } = initialConfig || {};
  const onBlockLoad = useCallback(() => {
    setLoaded(true);
  }, []);

  const initWithAutomation = useCallback(async () => {
    if (!user || !page || !page.workspaceId || !loaded) return;
    try {
      const newConfig = await api.callAutomation(page.workspaceId, automation);
      setConfig((prev = {}) => ({ ...prev, ...newConfig }));
    } catch {}
  }, [automation, loaded, page, user]);

  useEffect(() => {
    if (!user || !loaded || !events) return;
    onLoad && onLoad();
    // Set listeners
    let off: Function[] = [];
    if (updateOn) {
      off.push(
        events.on(updateOn, ({ payload: config }) => {
          setConfig((prev = {}) => ({ ...prev, ...config }));
          if (config.userTopics) {
            events.listenTopics({ event: updateOn, topics: config.userTopics });
          }
        })
      );
    }
    setListening(true);

    if (automation) {
      initWithAutomation();
    }

    return () => {
      off.forEach((off) => off());
    };
  }, [
    events,
    onLoad,
    updateOn,
    onInit,
    page,
    initialConfig,
    loaded,
    automation,
    initWithAutomation,
    user,
  ]);

  const alreadySentInit = useRef(false);
  useEffect(() => {
    if (!user || !listening || !events || alreadySentInit.current || !onInit)
      return;
    alreadySentInit.current = true;
    const payload: any = {
      page: page && page.id,
      config: initialConfig,
    };
    if (window.location.search) {
      payload.query = Array.from(
        new URLSearchParams(window.location.search).entries()
      ).reduce(
        (prev, [key, value]) => ({
          ...prev,
          [key]: value,
        }),
        {}
      );
    }
    events.emit(onInit, payload);
  }, [events, initialConfig, listening, onInit, page, user]);

  const onAppConfigUpdate = useCallback(
    async (newConfig: any) => {
      lock.current = true;
      setAppConfig(() => newConfig);
      if (name.match(/^http/) || !name.match(/\./)) return;
      const [appInstance] = name.split(/\./);
      if (!page?.workspaceId || !appInstance) return;
      return api.updateAppConfig(page.workspaceId, appInstance, newConfig);
    },
    [name, page]
  );

  if (!page) return null;

  return (
    <BLoader
      name={getBlockName(name)}
      url={url}
      appConfig={appConfig}
      onAppConfigUpdate={onAppConfigUpdate}
      api={api}
      language={language}
      workspaceId={`${page.workspaceId}`}
      events={events}
      config={config}
      layout={{
        container,
      }}
      onLoad={onBlockLoad}
    />
  );
};

export default BlockLoader;
