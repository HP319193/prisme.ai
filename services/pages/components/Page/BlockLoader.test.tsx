import renderer, { act } from 'react-test-renderer';
import BlockLoader from './BlockLoader';
import { usePage } from './PageProvider';

window.Prisme = {
  ai: {
    debug: {},
  },
} as any;

jest.mock('./PageProvider', () => {
  const mock = {
    page: {
      workspaceId: '42',
      appInstances: [
        {
          slug: 'Foo',
          blocks: {
            'Foo.Block1': 'http://fooblock1',
            'Foo.Block2': 'http://fooblock2',
          },
        },
      ],
    },
    events: {
      emit: jest.fn(),
    },
  };
  return {
    usePage: () => mock,
  };
});

jest.mock('../../../console/components/UserProvider', () => {
  const user = {};
  return {
    useUser() {
      return {
        user,
      };
    },
  };
});

it('should render a builtin block', () => {
  const root = renderer.create(<BlockLoader name="RichText" />);
  const child = root.root.children[0] as renderer.ReactTestInstance;
  expect(child.props.name).toBe('RichText');
  expect(child.props.url).toBe('');
  expect(child.props.language).toBe('en');
  expect(child.props.workspaceId).toBe('42');
  expect(child.props.config).not.toBeDefined();
});

it('should render an App block', () => {
  const { page } = usePage();
  const root = renderer.create(<BlockLoader name="Foo.Block1" />);

  act(() => {
    return;
  });

  const child = root.root.children[0] as renderer.ReactTestInstance;
  expect(child.props.name).toBe('Foo.Block1');
  expect(child.props.url).toBe('http://fooblock1');
  expect(child.props.language).toBe('en');
  expect(child.props.workspaceId).toBe('42');
  expect(child.props.config).not.toBeDefined();
});

it('should render a custom block with url', () => {
  const root = renderer.create(<BlockLoader name="http://custom" />);

  act(() => {
    return;
  });

  const child = root.root.children[0] as renderer.ReactTestInstance;
  expect(child.props.name).toBe('http://custom');
  expect(child.props.url).toBe('http://custom');
  expect(child.props.language).toBe('en');
  expect(child.props.workspaceId).toBe('42');
  expect(child.props.config).not.toBeDefined();
});

it('should render an uninstalled App block', () => {
  console.error = jest.fn();
  const root = renderer.create(<BlockLoader name="Bar.Block1" />);

  act(() => {
    return;
  });

  expect(console.error).toHaveBeenCalledWith(
    `"Bar.Block1" Block is not installed`
  );
});

it('should init block', () => {
  const { events } = usePage();
  const config = {
    onInit: 'initBlock',
  };
  const onLoad = jest.fn();
  const root = renderer.create(
    <BlockLoader name="Foo.Block1" config={config} onLoad={onLoad} />
  );

  act(() => {
    return;
  });

  const child = root.root.children[0] as renderer.ReactTestInstance;

  act(() => {
    child.props.onLoad();
  });

  expect(events!.emit).toHaveBeenCalledWith('initBlock', {
    config: { onInit: 'initBlock' },
    language: 'en',
    page: undefined,
  });
  expect(onLoad).toHaveBeenCalled();
});
