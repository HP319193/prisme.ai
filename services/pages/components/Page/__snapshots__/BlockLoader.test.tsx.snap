// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render a builtin block 1`] = `null`;
